### PR TITLE
Pattern-2

### PR DESCRIPTION
**Pull request by BLasan**
**Changes**

- Created Pattern-2 Deployment for APIM
- Introduced synchronization for artifacts
- Removed pvc
- Added scenario-10, scenario-11 & scenario-12 for pattern-2

### CODE REVIEW

File: artifacts/api-manager-artifacts/pattern-2/wso2am-p2-analytics-dash-bin.yaml [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: wso2am-p2-analytics-dash-bin
+  namespace : wso2-system
+data:
```
Comment (pubudu538) [inline artifacts/api-manager-artifacts/pattern-2/wso2am-p2-analytics-dash-bin.yaml]: Shall we use common analytics for all the patterns? Basically we can use single analytics resources for all the patterns.

File: artifacts/operator-configs/controller-conf.yaml [hunk 2]
```diff
@@ -55,14 +55,47 @@ data:
...
+  apim-analytics-deployment-securityContext-runAsUser: "802"
 
   # pattern specific configs for apim-instances
   p1-apim-deployment-image: "wso2/wso2am:3.2.0"
-  p1-apim-deployment-resources-requests-cpu: "2000m"
-  p1-apim-deployment-resources-requests-memory: "2Gi"
+  p1-apim-deployment-resources-requests-cpu: "1000m"
+  p1-apim-deployment-resources-requests-memory: "1Gi"
   p1-apim-deployment-resources-limits-cpu: "2000m"
   p1-apim-deployment-resources-limits-memory: "2Gi"
```
Comment (pubudu538) [inline artifacts/operator-configs/controller-conf.yaml:70]: Let's use some of the following as common values.

1. Images APIM and Analytics
2. Cpu and memory values
3. JVM heap values
4. Liveness and readiness probes

File: pkg/controller/pattern2/constants.go [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package pattern2
```
Comment (pubudu538) [inline pkg/controller/pattern2/constants.go:21]: License headers

File: pkg/controller/pattern2/volumes.go [hunk 4]
```diff
@@ -0,0 +1,552 @@
...
+		// 	devpubtm2volume = append(devpubtm2volume, corev1.Volume{
+		// 		Name: c.Name,
+		// 		VolumeSource: corev1.VolumeSource{
+		// 			PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
+		// 				ClaimName: c.Name,
+		// 			},
+		// 		},
+		// 	})
+		// }
+	}
```
Comment (pubudu538) [inline pkg/controller/pattern2/volumes.go:123]: Let's remove commented code.

File: scenarios/scenario-10/wso2-apim.yaml [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+# specific language governing permissions and limitations
+# under the License.
+
+apiVersion: apim.wso2.com/v1alpha1
+kind: APIManager
+metadata:
+  name: apim-pattern-2
+spec:
+  pattern: Pattern-2
+  enableAnalytics: 'true'
```
Comment (pubudu538) [inline scenarios/scenario-10/wso2-apim.yaml]: By default we can set enableAnalytics to true and not need to define this property here. It should be an optional property.

**Review by pubudu538 — CHANGES_REQUESTED** [review]

File: pkg/controller/pattern2/volumes.go (hunk 4 above)
Comment (BLasan) [reply pkg/controller/pattern2/volumes.go:123]: Removed. Please review

File: artifacts/api-manager-artifacts/pattern-2/wso2am-p2-analytics-dash-bin.yaml (hunk 1 above)
Comment (BLasan) [reply artifacts/api-manager-artifacts/pattern-2/wso2am-p2-analytics-dash-bin.yaml]: only analytics-bin configs are same. But have different configs in worker & dashboard analytics

File: pkg/controller/controller.go [hunk 6]
```diff
@@ -840,6 +845,12 @@ func (c *Controller) syncHandler(key string) error {
 
 	}
 
+	klog.Info("Phase 5 Sucess")
```
Comment (pubudu538) [inline pkg/controller/controller.go]: Remove unwanted logs.

**Review by pubudu538 — APPROVED** [review]

**Review by pubudu538 — APPROVED** [review]

File: artifacts/api-manager-artifacts/pattern-2/wso2am-p2-mysql-dbscripts.yaml [hunk 7]
```diff
@@ -0,0 +1,2491 @@
...
+
+    USE APIM_ANALYTICS_DB;
+
+    CREATE TABLE IF NOT EXISTS AM_USAGE_UPLOADED_FILES (
+    FILE_NAME varchar(255) NOT NULL,
+    FILE_TIMESTAMP TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
+    FILE_PROCESSED tinyint(1) DEFAULT 0,
+    FILE_CONTENT MEDIUMBLOB DEFAULT NULL,
+    PRIMARY KEY (FILE_NAME, FILE_TIMESTAMP)
+    );
```
Comment (senthuran16) [inline artifacts/api-manager-artifacts/pattern-2/wso2am-p2-mysql-dbscripts.yaml:64]: Shall we indent statements similar to this?
eg:
```sql
    CREATE TABLE IF NOT EXISTS AM_USAGE_UPLOADED_FILES (
        FILE_NAME varchar(255) NOT NULL,
        FILE_TIMESTAMP TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
        FILE_PROCESSED tinyint(1) DEFAULT 0,
        FILE_CONTENT MEDIUMBLOB DEFAULT NULL,
        PRIMARY KEY (FILE_NAME, FILE_TIMESTAMP)
    );
```

File: pkg/controller/pattern2/service.go [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+			},
+		},
+		Spec: corev1.ServiceSpec{
+			Selector: labels,
+			Ports:    workerports,
+		},
+	}
+}
+
+//WorkerHeadlessService is for...
```
Comment (senthuran16) [inline pkg/controller/pattern2/service.go:165]: Seems this comment is incomplete?

File: artifacts/api-manager-artifacts/pattern-2/wso2am-p2-analytics-bin.yaml [hunk 9]
```diff
@@ -0,0 +1,330 @@
+# Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (GDLMadushanka) [inline artifacts/api-manager-artifacts/pattern-2/wso2am-p2-analytics-bin.yaml:1]: shall we change to 2022 and LLC

File: pkg/controller/controller.go [hunk 10]
```diff
@@ -1199,3 +1205,664 @@ func (c *Controller) handleObject(obj interface{}) {
...
+		}
+	}
+
+	// Get mysql deployment name using hardcoded value
+	mysqldeployment, err := c.deploymentsLister.Deployments(apimanager.Namespace).Get(mysqldeploymentName)
+
+	if useMysqlPod {
+		// If the resource doesn't exist, we'll create it
+		if errors.IsNotFound(err) {
+			//y:= pattern1.AssignMysqlConfigMapValues(apimanager,configmap)
```
Comment (GDLMadushanka) [inline pkg/controller/controller.go:1390]: sall we remove the comment if not important

File: pkg/controller/controller.go [hunk 11]
```diff
@@ -1199,3 +1205,664 @@ func (c *Controller) handleObject(obj interface{}) {
...
+	// Get analytics worker deployment name using hardcoded value
+
+	// workerdeployment, err := c.deploymentsLister.Deployments(apimanager.Namespace).Get(workerDeploymentName)
+	workerdeployment, err := c.statefulSetsLister.StatefulSets(apimanager.Namespace).Get(workerDeploymentName)
+	if enableAnalytics {
+		// If the resource doesn't exist, we'll create it
+		if errors.IsNotFound(err) {
+			y := pattern2.AssignApimAnalyticsWorkerConfigMapValues(apimanager, configmap, worknum)
+
+			// workerdeployment, err = c.kubeclientset.AppsV1().Deployments(apimanager.Namespace).Create(pattern1.WorkerDeployment(apimanager, y, worknum))
```
Comment (GDLMadushanka) [inline pkg/controller/controller.go:1455]: remove the comment and use a meaningfull variable name for y

Comment (BLasan) [reply pkg/controller/controller.go:1455]: addressed via https://github.com/wso2/k8s-wso2am-operator/pull/89/commits/788038d4b86839a0ede2e666405f0c4f501c0dbf

File: pkg/controller/controller.go (hunk 10 above)
Comment (BLasan) [reply pkg/controller/controller.go:1390]: addressed via https://github.com/wso2/k8s-wso2am-operator/pull/89/commits/788038d4b86839a0ede2e666405f0c4f501c0dbf

File: pkg/controller/pattern2/service.go (hunk 8 above)
Comment (BLasan) [reply pkg/controller/pattern2/service.go:165]: addressed via https://github.com/wso2/k8s-wso2am-operator/pull/89/commits/788038d4b86839a0ede2e666405f0c4f501c0dbf

File: artifacts/api-manager-artifacts/pattern-2/wso2am-p2-mysql-dbscripts.yaml (hunk 7 above)
Comment (BLasan) [reply artifacts/api-manager-artifacts/pattern-2/wso2am-p2-mysql-dbscripts.yaml:64]: addressed via https://github.com/wso2/k8s-wso2am-operator/pull/89/commits/8b163758448b29fec66bc8139ad184f405bf1f28

**Pull request merged**